epochStart/shardchain/trigger.go

```diff
@@ -181,14 +190,69 @@ func NewEpochStartTrigger(args *ArgsShardEpochStartTrigger) (*trigger, error) {
 		appStatusHandler:            &statusHandler.NilStatusHandler{},
 	}
 
-	newTrigger.headersPool.RegisterHandler(newTrigger.receivedMetaBlock)
+	t.headersPool.RegisterHandler(t.receivedMetaBlock)
 
-	err := newTrigger.saveState(newTrigger.triggerStateKey)
+	err := t.saveState(t.triggerStateKey)
 	if err != nil {
 		return nil, err
 	}
 
-	return newTrigger, nil
+	t.mapMissingMiniblocks = make(map[string]uint32)
+	go t.requestMissingMiniblocks()
+
+	return t, nil
+}
+
+func (t *trigger) clearMissingMiniblocksMap(epoch uint32) {
+	t.mutMissingMiniblocks.Lock()
+	defer t.mutMissingMiniblocks.Unlock()
+
+	for hash, epochOfMissingMb := range t.mapMissingMiniblocks {
+		if epochOfMissingMb <= epoch {
+			delete(t.mapMissingMiniblocks, hash)
+		}
+	}
+}
+
+func (t *trigger) requestMissingMiniblocks() {
+	for {
+		time.Sleep(sleepTime)
+
+		t.mutMissingMiniblocks.RLock()
+		if len(t.mapMissingMiniblocks) == 0 {
+			t.mutMissingMiniblocks.RUnlock()
+			continue
+		}
+
+		missingMiniblocks := make([][]byte, len(t.mapMissingMiniblocks))
+		for hash := range t.mapMissingMiniblocks {
+			missingMiniblocks = append(missingMiniblocks, []byte(hash))
+			log.Debug("trigger.requestMissingMiniblocks", "hash", []byte(hash))
+		}
+		t.mutMissingMiniblocks.RUnlock()
+
+		go t.requestHandler.RequestMiniBlocks(core.MetachainShardId, missingMiniblocks)
```

Comment:
this is done also in the validatorInfoProcessor; is it really needed in both places?

Reply:
Actually the problem was that if after 1 second of waiting in validatorInfoProcessor some of miniblocks are still missing there was no mechanism to request them again and the node would be stuck to this point. Now, in trigger if some missing hashes are returned they would be requested on this go routine and when they all are received the method updateTrigger... would be called again